File: contrib/scripts/plot_speedtest.py
#!/usr/bin/env python3

# This script renders an afp_speedtest CSV capture into a throughput-vs-size
# chart (SVG + PNG) for publishing in CI, mirroring the flamegraph workflow.
#
# (c) 2026 Andy Lemin (@andylemin)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

"""Render afp_speedtest CSV output as a throughput-vs-file-size chart.

The speedtest tool (test/testsuite/speedtest.c) emits, when run with -c, one
"File Size Performance Summary" block per operation:

    # File Size Performance Summary for Read
    file_size_bytes,file_size_mb,mean_throughput_mbs,median_throughput_mbs,min_throughput_mbs,max_throughput_mbs,stddev_ms,mean_time_ms
    4096,0.004,66.15,65.65,52.79,93.01,0.0,0.1
    ...

This parser keys on those headers and ignores any surrounding log noise (the
container entrypoint prints setup messages on the same stdout), so the raw
`docker run ... > speed.csv` capture can be fed in directly.

Output: a single plot with X = file size (log2 scale) and Y = median
throughput (MB/s), one line per operation, with a shaded min..max band.

Usage:
    plot_speedtest.py INPUT.csv [-o OUTPUT_BASENAME] [--title TITLE]
                      [--subtitle SUBTITLE] [--baseline BASELINE.csv]

Writes OUTPUT_BASENAME.png (default basename: "speedtest").
"""

import argparse
import re
import sys

import matplotlib

matplotlib.use("Agg")  # headless / no display, for CI runners
import matplotlib.pyplot as plt
from matplotlib.ticker import FuncFormatter

# Operations in the order the speedtest runs them, with stable colours so the
# chart reads the same across commits.
OPERATIONS = ["Read", "Write", "Copy", "ServerCopy"]
OP_COLORS = {
    "Read": "#1f77b4",        # blue
    "Write": "#d62728",       # red
    "Copy": "#2ca02c",        # green
    "ServerCopy": "#ff7f0e",  # orange
}

SUMMARY_HEADER_RE = re.compile(
    r"#\s*File Size Performance Summary for\s+(\S+)", re.IGNORECASE
)
# Column header line that immediately precedes the data rows.
COLUMN_HEADER_PREFIX = "file_size_bytes,"
# Optional self-describing config line the speedtest may emit in CSV mode,
# e.g. "# Config: quantum_kb=1024,iterations=5,warmup=1,afp=3.4".
CONFIG_LINE_RE = re.compile(r"#\s*Config:\s*(.+)", re.IGNORECASE)
# Per-iteration data row: "Op,iteration,file_size_mb,microseconds,throughput".
ITER_ROW_RE = re.compile(r"^([A-Za-z]+),(\d+),([\d.]+),(\d+),([\d.]+)\s*$")


def _parse_config_meta(line):
    """Return the k=v pairs from a "# Config: k=v,k=v" line, else None."""
    cfg = CONFIG_LINE_RE.match(line)
    if not cfg:
        return None
    pairs = {}
    for pair in cfg.group(1).split(","):
        if "=" in pair:
            k, v = pair.split("=", 1)
            pairs[k.strip()] = v.strip()
    return pairs


def _parse_summary_row(line):
    """Parse a summary data row into a row dict, or None if it isn't one.

    Row layout: size_bytes,size_mb,mean,median,min,max,stddev,mean_time.
    """
    parts = line.split(",")
    if len(parts) < 6:
        return None
    try:
        return {
            "size_bytes": float(parts[0]),
            "mean": float(parts[2]),
            "median": float(parts[3]),
            "min": float(parts[4]),
            "max": float(parts[5]),
        }
    except ValueError:
        return None


def parse_csv(path):
    """Parse a speedtest CSV capture.

    Returns (results, meta) where:
      * results = {operation: [row_dict, ...]}, each row_dict with float keys
        size_bytes, mean, median, min, max; rows sorted by ascending size.
      * meta = dict of run metadata derived from the capture: "iterations"
        (max per-iteration index seen) plus any key=value pairs from an
        optional "# Config: k=v,k=v" line emitted by the tool.
    """
    results = {}
    meta = {}
    max_iter = 0
    current_op = None
    in_data = False

    with open(path, "r", encoding="utf-8", errors="replace") as fh:
        for raw in fh:
            line = raw.strip()

            cfg = _parse_config_meta(line)
            if cfg is not None:
                meta.update(cfg)
                continue

            it = ITER_ROW_RE.match(line)
            if it:
                max_iter = max(max_iter, int(it.group(2)))

            header = SUMMARY_HEADER_RE.search(line)
            if header:
                current_op = header.group(1)
                results.setdefault(current_op, [])
                in_data = False
                continue

            if current_op is None:
                continue

            if line.startswith(COLUMN_HEADER_PREFIX):
                in_data = True
                continue

            if not in_data:
                continue

            row = _parse_summary_row(line)
            if row is None:
                in_data = False  # blank line / next section / log noise
            else:
                results[current_op].append(row)

    cleaned = {
        op: sorted(rows, key=lambda r: r["size_bytes"])
        for op, rows in results.items() if rows
    }
    if max_iter and "iterations" not in meta:
        meta["iterations"] = str(max_iter)
    return cleaned, meta


def format_size(size_bytes, _pos=None):
    """Format an exact byte count as an integer KiB/MiB axis label."""
    kib = size_bytes / 1024.0
    if kib < 1024.0:
        return f"{int(round(kib))}K"
    return f"{int(round(kib / 1024.0))}M"


def build_info_lines(results, meta):
    """Compose the run-config lines shown in the chart's info box.

    Values come from the CSV where available (size range, sweep points,
    iterations, files-per-op) and from --meta key=value pairs the caller
    injects for things the CSV cannot carry (quantum, ARC dircache, AFP
    version). Missing values are simply omitted rather than guessed.
    """
    ops = [op for op in OPERATIONS if op in results]
    ops += [op for op in results if op not in OPERATIONS]
    # Size range / point count from any populated operation.
    sizes = []
    for rows in results.values():
        sizes = [r["size_bytes"] for r in rows]
        if sizes:
            break

    lines = []
    if sizes:
        lines.append(f"Size sweep: {format_size(min(sizes))}–{format_size(max(sizes))}")
    if "iterations" in meta:
        warm = f" +{meta['warmup']} warmup" if meta.get("warmup") else ""
        lines.append(f"Iterations: {meta['iterations']}{warm} per point")
    lines.append(f"Ops: {', '.join(ops)}")
    if meta.get("afp"):
        lines.append(f"AFP: {meta['afp']}")
    if meta.get("quantum_kb"):
        lines.append(f"Quantum: {meta['quantum_kb']} KB")
    dc = []
    if meta.get("dircache_mode"):
        dc.append(meta["dircache_mode"].upper())
    if meta.get("dircache_size"):
        dc.append(f"size {meta['dircache_size']}")
    if meta.get("dircache_validation_freq"):
        dc.append(f"freq {meta['dircache_validation_freq']}")
    if dc:
        lines.append("Dircache: " + " · ".join(dc))
    return lines


# Background grey matching the FlameGraph SVG palette (RGB 238,238,238).
BACKGROUND_COLOR = "#EEEEEE"


def _draw_operation(ax, op, rows, baseline):
    """Plot one operation's median line, min–max band, and mean reference."""
    color = OP_COLORS.get(op, None)
    sizes = [r["size_bytes"] for r in rows]
    median = [r["median"] for r in rows]
    ax.plot(sizes, median, marker="o", markersize=4, linewidth=2,
            color=color, label=op, zorder=3)
    # min..max band gives a sense of run-to-run variance.
    ax.fill_between(sizes, [r["min"] for r in rows], [r["max"] for r in rows],
                    color=color, alpha=0.12, zorder=1)

    # Horizontal reference at this op's mean throughput, with a value label
    # just right of the Y axis so it never overlaps the Y tick labels.
    avg = sum(median) / len(median)
    ax.axhline(avg, color=color, linestyle="--", linewidth=1.0, alpha=0.6,
               zorder=2, label=f"{op} mean ({avg:.0f} MB/s)")
    ax.text(0.012, avg, f"{avg:.0f}", transform=ax.get_yaxis_transform(),
            ha="left", va="bottom", fontsize=7, color=color,
            fontweight="bold", zorder=5,
            bbox=dict(boxstyle="round,pad=0.15", facecolor="white",
                      edgecolor="none", alpha=0.7))

    if baseline and op in baseline:
        b = sorted(baseline[op], key=lambda r: r["size_bytes"])
        ax.plot([r["size_bytes"] for r in b], [r["median"] for r in b],
                linestyle="--", linewidth=1.3, color=color, alpha=0.6,
                zorder=2, label=f"{op} (baseline)")


def _configure_xaxis(ax, results):
    """Log2 X axis with one integer KiB/MiB tick per swept size, clamped."""
    ax.set_xscale("log", base=2)
    all_sizes = sorted({r["size_bytes"] for rows in results.values() for r in rows})
    if all_sizes:
        ax.set_xticks(all_sizes)
        ax.set_xticklabels([format_size(s) for s in all_sizes], fontsize=8)
        ax.minorticks_off()
        ax.set_xlim(all_sizes[0], all_sizes[-1])
    ax.xaxis.set_major_formatter(FuncFormatter(format_size))


def plot(results, out_base, title, subtitle, baseline=None, meta=None):
    """Render the throughput-vs-size chart to <out_base>.png."""
    meta = meta or {}
    fig, ax = plt.subplots(figsize=(11, 6.5))
    fig.patch.set_facecolor(BACKGROUND_COLOR)
    ax.set_facecolor(BACKGROUND_COLOR)

    # Known operations first (stable colour/order), then any unexpected ones.
    ordered = [op for op in OPERATIONS if results.get(op)]
    ordered += [op for op in results if op not in OPERATIONS and results[op]]
    for op in ordered:
        _draw_operation(ax, op, results[op], baseline)

    if not ordered:
        sys.stderr.write(
            "ERROR: no speedtest summary data found in input — "
            "was afp_speedtest run with -c (CSV)?\n"
        )
        return False

    _configure_xaxis(ax, results)
    ax.set_xlabel("File size")
    ax.set_ylabel("Throughput (MB/s)")
    ax.grid(True, which="both", linestyle=":", alpha=0.4)
    legend = ax.legend(loc="upper left", fontsize=9, framealpha=0.9)
    legend.get_frame().set_facecolor("#F5F5F5")

    suptitle = title or "Netatalk AFP Speedtest — throughput vs file size"
    fig.suptitle(suptitle, fontsize=14, fontweight="bold", y=0.96)

    # Header text below the title: the caller's subtitle plus the run-config
    # lines, rendered in the title area rather than a floating box so it never
    # overlaps the curves or the bottom-right data corner.
    header_lines = []
    if subtitle:
        header_lines.append(subtitle)
    # Pack the config items into a single "·"-separated line under the
    # subtitle (compact, won't push the plot down or overlap the curves).
    info = build_info_lines(results, meta)
    if info:
        header_lines.append(" · ".join(info))
    if header_lines:
        ax.set_title("\n".join(header_lines), fontsize=8.5, color="#222222")

    fig.tight_layout(rect=(0, 0, 1, 0.97))

    png_path = f"{out_base}.png"
    # facecolor=fig... so savefig keeps the grey background (it defaults to white).
    fig.savefig(png_path, format="png", dpi=130, facecolor=fig.get_facecolor())
    plt.close(fig)
    sys.stderr.write(f"Wrote {png_path}\n")
    return True


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", help="speedtest CSV capture (afp_speedtest -c)")
    parser.add_argument(
        "-o", "--output", default="speedtest",
        help="output basename (default: speedtest -> speedtest.svg/.png)",
    )
    parser.add_argument("--title", default=None, help="chart title")
    parser.add_argument("--subtitle", default=None, help="chart subtitle")
    parser.add_argument(
        "--baseline", default=None,
        help="optional prior CSV to overlay as dashed baseline lines",
    )
    parser.add_argument(
        "--meta", action="append", default=[], metavar="KEY=VALUE",
        help="run-config metadata for the info box (repeatable). Known keys: "
             "afp, quantum_kb, dircache_mode, dircache_size, "
             "dircache_validation_freq, iterations, warmup. Overrides values "
             "parsed from the CSV.",
    )
    args = parser.parse_args(argv)

    results, meta = parse_csv(args.input)
    # CLI --meta overrides anything parsed from the CSV.
    for pair in args.meta:
        if "=" in pair:
            k, v = pair.split("=", 1)
            meta[k.strip()] = v.strip()

    baseline = None
    if args.baseline:
        try:
            baseline, _ = parse_csv(args.baseline)
        except OSError as exc:
            sys.stderr.write(f"WARNING: could not read baseline: {exc}\n")

    ok = plot(results, args.output, args.title, args.subtitle, baseline, meta)
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: contrib/scripts/plot_speedtest.pl
#!/usr/bin/env perl

# This script renders an afp_speedtest CSV capture into a throughput-vs-size
# chart (PNG) for publishing in CI, mirroring the flamegraph workflow.
# Rendering is delegated to gnuplot (gnuplot-nox is sufficient), so only
# core Perl modules are used.
#
# (c) 2026 Andy Lemin (@andylemin)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# The speedtest tool (test/testsuite/speedtest.c) emits, when run with -c,
# one "File Size Performance Summary" block per operation:
#
#     # File Size Performance Summary for Read
#     file_size_bytes,file_size_mb,mean_throughput_mbs,median_throughput_mbs,min_throughput_mbs,max_throughput_mbs,stddev_ms,mean_time_ms
#     4096,0.004,66.15,65.65,52.79,93.01,0.0,0.1
#     ...
#
# This parser keys on those headers and ignores any surrounding log noise
# (the container entrypoint prints setup messages on the same stdout), so
# the raw `docker run ... > speed.csv` capture can be fed in directly.
#
# Output: a single plot with X = file size (log2 scale) and Y = mean
# throughput (MB/s), one line per operation, with a shaded min..max band.
#
# Usage:
#     plot_speedtest.pl INPUT.csv [-o OUTPUT_BASENAME] [--title TITLE]
#                       [--subtitle SUBTITLE] [--baseline BASELINE.csv]
#                       [--meta KEY=VALUE ...]
#
# Writes OUTPUT_BASENAME.png (default basename: "speedtest").

use strict;
use warnings;
use Getopt::Long qw(GetOptions);

# Operations in the order the speedtest runs them, with stable colors so
# the chart reads the same across commits.
my @OPERATIONS = qw(Read Write Copy ServerCopy);
my %OP_COLORS = (
                 Read       => '#1f77b4',    # blue
                 Write      => '#d62728',    # red
                 Copy       => '#2ca02c',    # green
                 ServerCopy => '#ff7f0e',    # orange
);
my $DEFAULT_COLOR = '#7f7f7f';               # any unexpected operation

# Near-white canvas (RGB 246,248,250) for a clean modern background.
my $BACKGROUND_COLOR = '#F6F8FA';

# 11 x 5.2 inches at 130 dpi, matching the former matplotlib renderer.
my $WIDTH_PX  = 1430;
my $HEIGHT_PX = 676;

# Column header line that immediately precedes the data rows.
my $COLUMN_HEADER_PREFIX = 'file_size_bytes,';

# Parse a speedtest CSV capture into (results, meta): results = {op => [rows
# sorted by size]}; meta = run metadata (iterations from the max per-iteration
# index seen, plus any key=value pairs from an optional "# Config:" line).
sub parse_csv {
    my ($path) = @_;
    my (%results, %meta);
    my $max_iter   = 0;
    my $current_op = undef;
    my $in_data    = 0;

    open my $fh, '<', $path or die "cannot open $path: $!\n";
    while (my $line = <$fh>) {
        $line =~ s/^\s+|\s+$//g;

        # Optional self-describing config line, e.g.
        # "# Config: quantum_kb=1024,iterations=5,warmup=1,afp=3.4".
        if ($line =~ /#\s*Config:\s*(.+)/i) {
            for my $pair (split /,/, $1) {
                next unless index($pair, '=') >= 0;
                my ($k, $v) = split /=/, $pair, 2;
                s/^\s+|\s+$//g for ($k, $v);
                $meta{$k} = $v;
            }
            next;
        }

        # Per-iteration row: "Op,iteration,file_size_mb,microseconds,tput".
        if ($line =~ /^[A-Za-z]+,(\d+),[\d.]+,\d+,[\d.]+\s*$/) {
            $max_iter = $1 if $1 > $max_iter;
        }

        if ($line =~ /#\s*File Size Performance Summary for\s+(\S+)/i) {
            $current_op = $1;
            $results{$current_op} ||= [];
            $in_data = 0;
            next;
        }
        next unless defined $current_op;

        if (index($line, $COLUMN_HEADER_PREFIX) == 0) {
            $in_data = 1;
            next;
        }
        next unless $in_data;

        # Summary row: size_bytes,size_mb,mean,median,min,max,stddev,time.
        my @parts = split /,/, $line;
        if (@parts >= 6
             && !grep { !/^-?[\d.]+$/ } @parts[0, 2 .. 5]) {
            push @{$results{$current_op}}, {
                                            size_bytes => $parts[0] + 0,
                                            mean       => $parts[2] + 0,
                                            median     => $parts[3] + 0,
                                            min        => $parts[4] + 0,
                                            max        => $parts[5] + 0,
            };
        } else {
            $in_data = 0;    # blank line / next section / log noise
        }
    }
    close $fh;

    for my $op (keys %results) {
        if (!@{$results{$op}}) {
            delete $results{$op};
            next;
        }
        $results{$op} = [sort { $a->{size_bytes} <=> $b->{size_bytes} } @{$results{$op}}];
    }
    $meta{iterations} = $max_iter if $max_iter && !$meta{iterations};
    return (\%results, \%meta);
}

# Format an exact byte count as an integer KiB/MiB axis label.
sub format_size {
    my ($size_bytes) = @_;
    my $kib = $size_bytes / 1024.0;
    return sprintf('%dK', $kib + 0.5) if $kib < 1024.0;
    return sprintf('%dM', $kib / 1024.0 + 0.5);
}

# Known operations first (stable color/order), then any unexpected ones.
sub ordered_ops {
    my ($results) = @_;
    my @ops = grep { $results->{$_} } @OPERATIONS;
    push @ops, grep {
        my $op = $_;
        !grep { $_ eq $op } @OPERATIONS
    } sort keys %$results;
    return @ops;
}

# Run-config lines shown in the chart's info box, from the CSV and --meta.
# Missing values are omitted rather than guessed. Size sweep is omitted (the
# X axis already shows the range).
sub build_info_lines {
    my ($results, $meta) = @_;
    my @ops = ordered_ops($results);
    my @lines;
    if ($meta->{iterations}) {
        my $warm = $meta->{warmup} ? " +$meta->{warmup} warmup" : '';
        push @lines, "Iterations: $meta->{iterations}$warm";
    }
    push @lines, 'Ops: ' . join(', ', @ops);
    push @lines, "AFP: $meta->{afp}"               if $meta->{afp};
    push @lines, "Quantum: $meta->{quantum_kb} KB" if $meta->{quantum_kb};
    my @dc;
    push @dc, uc $meta->{dircache_mode}     if $meta->{dircache_mode};
    push @dc, "size $meta->{dircache_size}" if $meta->{dircache_size};
    push @dc, "freq $meta->{dircache_validation_freq}"
      if $meta->{dircache_validation_freq};
    push @lines, 'Dircache: ' . join(' · ', @dc) if @dc;
    return @lines;
}

# Escape a string for a double-quoted gnuplot string (newline -> \n escape).
sub gp_quote {
    my ($s) = @_;
    $s =~ s/\\/\\\\/g;
    $s =~ s/"/\\"/g;
    $s =~ s/\n/\\n/g;
    return $s;
}

sub plot {
    my ($results, $out_base, $title, $subtitle, $baseline, $meta) = @_;
    my @ordered = ordered_ops($results);
    if (!@ordered) {
        print STDERR "ERROR: no speedtest summary data found in input — " . "was afp_speedtest run with -c (CSV)?\n";
        return 0;
    }

    # Collect plot commands into separate z-layers (see @plots below) so a
    # later op's fill or line never paints over an earlier op's curve. Draw
    # order then no longer matches legend order, so every data plot is
    # notitle and the legend is rebuilt from keyentry items in per-op order.
    my $blocks = '';
    my (
         @bands, @means, @baselines, @meancurves, @maxes, @keyentries,
         @labels
    );
    my $idx = 0;
    for my $op (@ordered) {
        my $rows  = $results->{$op};
        my $color = $OP_COLORS{$op} || $DEFAULT_COLOR;
        my $tag   = "OP$idx";
        $blocks .= "\$$tag << EOD\n";
        for my $r (@$rows) {
            $blocks .= "$r->{size_bytes} $r->{mean} $r->{min} $r->{max}\n";
        }
        $blocks .= "EOD\n";

        # min..max band shows run-to-run variance.
        push @bands,
          "\$$tag using 1:3:4 with filledcurves "
          . "fillcolor rgb \"$color\" fillstyle transparent solid 0.12 "
          . 'notitle';
        push @meancurves,
          "\$$tag using 1:2 with linespoints linewidth 2 pointtype 7 "
          . "pointsize 0.5 linecolor rgb \"$color\" notitle";
        # Max (best run) traces the band top, alpha-blended (#80 = 50%
        # transparent) so it reads lighter than the solid mean.
        my $light = '#80' . substr($color, 1);
        push @maxes,
          "\$$tag using 1:4 with lines dashtype (2,1) linewidth 1 " . "linecolor rgb \"$light\" notitle";

        # Two horizontal references per op: solid = mean of the per-size means
        # (pairs with the solid mean curve), dotted = mean of the maxes (pairs
        # with the dotted max hats).
        my ($sum_mean, $sum_max) = (0, 0);
        for my $r (@$rows) {
            $sum_mean += $r->{mean};
            $sum_max  += $r->{max};
        }
        my $avg     = $sum_mean / scalar(@$rows);
        my $avg_max = $sum_max / scalar(@$rows);
        push @means,
          sprintf(
                  '%s with lines linewidth 1 linecolor rgb "%s" notitle',
                  $avg, $color
          );
        push @means,
          sprintf(
                  '%s with lines dashtype (2,1) linewidth 1 linecolor rgb "%s" ' . 'notitle',
                  $avg_max, $light
          );
        # Value labels: avg-mean left-anchored at the Y axis, avg-max
        # right-anchored at the right edge, so the two never collide. "boxed"
        # gives a white backing (see set style textbox) for grid readability.
        push @labels,
          sprintf(
                    'set label %d "%.0f" at graph 0.012, first %s left front '
                  . 'offset 0,0.5 boxed font ",5" textcolor rgb "%s"',
                  $idx + 10, $avg, $avg, $color
          );
        push @labels,
          sprintf(
                    'set label %d "%.0f" at graph 0.988, first %s right front '
                  . 'offset 0,0.5 boxed font ",5" textcolor rgb "%s"',
                  $idx + 30, $avg_max, $avg_max, $color
          );

        # Legend: only the per-op mean curve; the avg-max/quantum overlays are
        # explained by the on-chart labels, keeping the key compact.
        push @keyentries,
          sprintf(
                    'keyentry with linespoints linewidth 2 pointtype 7 '
                  . 'pointsize 0.5 linecolor rgb "%s" title "%s (mean)"',
                  $color, gp_quote($op)
          );

        # Mark the dircache quantum (1 MiB) on the Read curve so the
        # rfork-cache boundary is visible.
        if ($op eq 'Read') {
            for my $r (@$rows) {
                next unless $r->{size_bytes} == 1024 * 1024;
                push @labels,
                  sprintf(
                            'set label 9 "quantum" at first %s, first %s '
                          . 'offset 0.6,0.8 front font ",5" '
                          . "textcolor rgb \"$color\"",
                          $r->{size_bytes}, $r->{mean}
                  );
                last;
            }
        }

        if ($baseline && $baseline->{$op} && @{$baseline->{$op}}) {
            my $btag = "BASE$idx";
            $blocks .= "\$$btag << EOD\n";
            for my $r (@{$baseline->{$op}}) {
                $blocks .= "$r->{size_bytes} $r->{mean}\n";
            }
            $blocks .= "EOD\n";
            push @baselines,
              "\$$btag using 1:2 with lines dashtype (4,2) linewidth 1.3 " . "linecolor rgb \"$color\" notitle";
            push @keyentries,
              "keyentry with lines dashtype (4,2) linewidth 1.3 "
              . "linecolor rgb \"$color\" "
              . "title \"@{[ gp_quote($op) ]} (baseline)\"";
        }
        $idx++;
    }
    # Bottom-to-top: bands then their max "hats" (the band top edge), baseline
    # overlays, mean curves, then the avg reference lines on top so they
    # stay visible across the data, then the (non-drawing) legend entries.
    my @plots = (
                 @bands, @maxes, @baselines, @meancurves, @means,
                 @keyentries
    );

    # Log2 X axis with one integer KiB/MiB tick per swept size, clamped.
    my %all_sizes;
    for my $rows (values %$results) {
        for my $r (@$rows) {
            $all_sizes{$r->{size_bytes}} = 1;
        }
    }
    my @all_sizes = sort { $a <=> $b } keys %all_sizes;
    my @tic_specs = map  { sprintf '"%s" %s', format_size($_), $_ } @all_sizes;
    my $tics      = join(', ', @tic_specs);

    my @header;
    push @header, $subtitle if defined $subtitle && $subtitle ne '';
    my @info = build_info_lines($results, $meta);
    push @header, join(' · ', @info) if @info;
    my $header = gp_quote(join("\n", @header));

    my $suptitle = gp_quote($title || 'Netatalk AFP Speedtest — throughput vs file size');
    my $png      = "$out_base.png";
    my $labels   = join("\n",          @labels);
    my $plots    = join(", \\\n     ", @plots);

    my $gp_bin = $ENV{GNUPLOT_BIN} || 'gnuplot';
    open my $gp, '|-', $gp_bin or die "cannot run $gp_bin: $!\n";
    print $gp <<"EOF";
# fontscale/linewidth/pointscale 1.8 matches matplotlib's dpi=130 (130/72),
# so the point sizes below mirror the old matplotlib values. DejaVu Sans is
# matplotlib's default; pangocairo falls back to system sans if it is absent.
set terminal pngcairo noenhanced size $WIDTH_PX,$HEIGHT_PX \\
    fontscale 1.8 linewidth 1.8 pointscale 1.8 \\
    background rgb "$BACKGROUND_COLOR" font "DejaVu Sans,7"
set output "@{[ gp_quote($png) ]}"

set border 15 linewidth 0.8 linecolor rgb "#333333"
set tics nomirror out scale 0.6

# Title block manually placed above the plot for a fixed title/subtitle gap.
set tmargin at screen 0.84
set bmargin at screen 0.12
set label 1 "$suptitle" at screen 0.5, screen 0.95 center \\
    font "DejaVu Sans:Bold,11"
set label 2 "$header" at screen 0.5, screen 0.885 center \\
    font "DejaVu Sans,5.5" textcolor rgb "#222222"

set logscale x 2
# 1.001 factor so the boundary tick label is not clipped.
set xrange [$all_sizes[0]:@{[ $all_sizes[-1] * 1.001 ]}]
set xtics font ",7" ($tics)
set xlabel "File size"
set ylabel "Throughput (MB/s)"
# Dotted grid; minor lines fainter so they don't read heavy.
set grid xtics ytics mxtics mytics linetype 1 dashtype (1,2) linewidth 1 \\
    linecolor rgb "#BBBBBB", linetype 1 dashtype (1,2) linewidth 1 \\
    linecolor rgb "#DDDDDD"
# Legend in the free top-left outer canvas, outside the plot area.
set key at screen 0.005, screen 0.995 top left Left reverse samplen 1.5 \\
    width -1 spacing 1.1 font ",6" \\
    box linewidth 0.6 linecolor rgb "#CCCCCC" opaque fillcolor rgb "#F5F5F5"
set style textbox opaque fillcolor rgb "#FFFFFF" noborder margins 0.6,0.6
# Column headers over the two value-label stacks (means across the sweep).
set label 3 "avg mean" at graph 0.012, graph 1.02 left front \\
    font "DejaVu Sans:Bold,6" textcolor rgb "#444444"
set label 4 "avg max" at graph 0.988, graph 1.02 right front \\
    font "DejaVu Sans:Bold,6" textcolor rgb "#444444"
$labels

$blocks
plot $plots
EOF
    close $gp or die "gnuplot failed (status $?)\n";
    print STDERR "Wrote $png\n";
    return 1;
}

sub main {
    my $output   = 'speedtest';
    my $title    = undef;
    my $subtitle = undef;
    my $baseline_path;
    my @meta_args;
    GetOptions(
               'o|output=s' => \$output,
               'title=s'    => \$title,
               'subtitle=s' => \$subtitle,
               'baseline=s' => \$baseline_path,
               'meta=s'     => \@meta_args,
      )
      or die "usage: plot_speedtest.pl INPUT.csv [-o BASENAME] "
      . "[--title T] [--subtitle S] [--baseline B.csv] "
      . "[--meta KEY=VALUE ...]\n";
    my $input = shift @ARGV
      or die "missing input CSV (afp_speedtest -c capture)\n";

    my ($results, $meta) = parse_csv($input);
    # CLI --meta overrides anything parsed from the CSV.
    for my $pair (@meta_args) {
        next unless index($pair, '=') >= 0;
        my ($k, $v) = split /=/, $pair, 2;
        s/^\s+|\s+$//g for ($k, $v);
        $meta->{$k} = $v;
    }

    my $baseline;
    if ($baseline_path) {
        if (-r $baseline_path) {
            ($baseline) = parse_csv($baseline_path);
        } else {
            print STDERR "WARNING: could not read baseline: $baseline_path\n";
        }
    }

    return plot($results, $output, $title, $subtitle, $baseline, $meta)
      ? 0
      : 1;
}

exit main();
